Libplanet.Headless/Hosting/LibplanetNodeService.cs

```diff
@@ -635,29 +637,53 @@ public override void Dispose()
             Log.Debug("Store disposed.");
         }
 
-        private string ResolvePluginPath(string path) =>
-            Uri.IsWellFormedUriString(path, UriKind.Absolute)
-                ? DownloadPlugin(path).Result
-                : path;
+        private string ResolvePluginPath(string path)
+        {
+            if (Uri.IsWellFormedUriString(path, UriKind.Absolute))
+            {
+                // Run the download on a background thread
+                return Task.Run(() => DownloadPlugin(path, SwarmCancellationToken)).GetAwaiter().GetResult();
+            }
 
-        private async Task<string> DownloadPlugin(string url)
+            return path;
+        }
+
+        private async Task<string> DownloadPlugin(string url, CancellationToken cancellationToken)
         {
-            var path = Path.Combine(Environment.CurrentDirectory, "plugins");
+            var basePath = Environment.GetEnvironmentVariable("PLUGIN_PATH") ?? Environment.CurrentDirectory;
+            var path = Path.Combine(basePath, "plugins");
             Directory.CreateDirectory(path);
-            var hashed = url.GetHashCode().ToString();
+            var hashed = CreateSha256Hash(url);
             var logger = Log.ForContext("LibplanetNodeService", hashed);
             using var httpClient = new HttpClient();
             var downloadPath = Path.Join(path, hashed + ".zip");
             var extractPath = Path.Join(path, hashed);
-            logger.Debug("Downloading...");
-            await File.WriteAllBytesAsync(
-                downloadPath,
-                await httpClient.GetByteArrayAsync(url, SwarmCancellationToken),
-                SwarmCancellationToken);
-            logger.Debug("Finished downloading.");
-            logger.Debug("Extracting...");
-            ZipFile.ExtractToDirectory(downloadPath, extractPath);
-            logger.Debug("Finished extracting.");
+            logger.Debug("Download Path: {downloadPath}", downloadPath);
+            logger.Debug("Extract Path: {extractPath}", extractPath);
+
+            if (!File.Exists(downloadPath))
+            {
+                logger.Debug("Downloading...");
+                var content = await httpClient.GetByteArrayAsync(url, cancellationToken);
+                await File.WriteAllBytesAsync(downloadPath, content, cancellationToken);
+                logger.Debug("Finished downloading.");
+            }
+            else
+            {
+                logger.Debug("Download skipped, file already exists.");
+            }
+
+            if (!Directory.Exists(extractPath))
```

Comment:
In the future, it needs to check checksum of directory's files. This is because if you unzip during a previous run and get hung up in the middle, you'll end up with an incomplete directory, and you can't check for this error just by looking at the existence of the directory.

Reply:
Could you check https://github.com/planetarium/NineChronicles.Headless/pull/2456/commits/2e06152c012211aa8a60d44da0d5169b1f42e34c if this would work? 🙏

Reply:
About this:

- If `checksum.txt` is in `extractPath`, the checksum which made of the files in `extractPath`, will be different. Because, when the first running, `checksum.txt = md5(filesInZip)` but when the second running, `checksum.txt = md5(filesInZip + checksum.txt)`. The checksum must be calculated from the provider side (e.g., GitHub Actions).
- Separately, we talked about checksums, but I think it's okay to simply use the completion flag after decompression. Of course, you still need a checksum to check if the DLL is corrupted or something, and as mentioned above, the provider should provide it.

Reply:
Let me know if https://github.com/planetarium/NineChronicles.Headless/commit/2e06152c012211aa8a60d44da0d5169b1f42e34c seems ok to you and I'll finalize this PR!